full-service/src/db/view_only_account.rs

```diff
@@ -0,0 +1,288 @@
+
+        let encoded_key = mc_util_serial::encode(view_private_key);
+
+        let new_view_only_account = NewViewOnlyAccount {
+            account_id_hex,
+            view_private_key: &encoded_key,
+            first_block_index,
+            // next block index will always be the same as
+            // first block index when importing an account
+            next_block_index: first_block_index,
+            import_block_index,
+            name,
+        };
+
+        diesel::insert_into(view_only_accounts::table)
+            .values(&new_view_only_account)
+            .execute(conn)?;
+
+        ViewOnlyAccount::get(account_id_hex, conn)
+    }
+
+    fn get(
+        account_id: &str,
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<ViewOnlyAccount, WalletDbError> {
+        use schema::view_only_accounts::dsl::{
+            account_id_hex as dsl_account_id, view_only_accounts,
+        };
+
+        match view_only_accounts
+            .filter((dsl_account_id).eq(account_id))
+            .get_result::<ViewOnlyAccount>(conn)
+        {
+            Ok(a) => Ok(a),
+            // Match on NotFound to get a more informative NotFound Error
+            Err(diesel::result::Error::NotFound) => Err(WalletDbError::AccountNotFound(
+                // TODO(cc) improve this error handling
+                // str::from_utf8(view_private_key)?,
+                "account not found".to_string(),
+            )),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn list_all(
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<Vec<ViewOnlyAccount>, WalletDbError> {
+        use schema::view_only_accounts;
+
+        Ok(view_only_accounts::table
+            .select(view_only_accounts::all_columns)
+            .load::<ViewOnlyAccount>(conn)?)
+    }
+
+    fn update_name(
+        &self,
+        new_name: &str,
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<(), WalletDbError> {
+        use schema::view_only_accounts::dsl::{
+            account_id_hex as dsl_account_id, name as dsl_name, view_only_accounts,
+        };
+
+        diesel::update(view_only_accounts.filter(dsl_account_id.eq(&self.account_id_hex)))
+            .set(dsl_name.eq(new_name))
+            .execute(conn)?;
+        Ok(())
+    }
+
+    fn update_next_block_index(
+        &self,
+        next_block_index: i64,
+        conn: &PooledConnection<ConnectionManager<SqliteConnection>>,
+    ) -> Result<(), WalletDbError> {
+        use schema::view_only_accounts::dsl::{
+            account_id_hex as dsl_account_id, next_block_index as dsl_next_block,
+            view_only_accounts,
+        };
+        diesel::update(view_only_accounts.filter(dsl_account_id.eq(&self.account_id_hex)))
+            .set(dsl_next_block.eq(next_block_index))
+            .execute(conn)?;
+        Ok(())
+    }
+
+    fn delete(
+        self,
```

Comment:
maybe newb question, but why pass `self` here rather than `&self`?

Reply:
using `self` here instead of `&self` because this is a delete method. I could be wrong, but I think it's unlikely that the entity will be used again after being deleted so it's okay for the method to take ownership